vaccel_args: Introduce helper functions for vAccel arguments

ilagomatis opened the pull request:
vaccel_args: Introduce helper functions for vAccel arguments

- genop: Add new field for argument type
- Add examples that cover ser/non-ser cases
- vaccel_args: Implement helpers
- vaccel_args: Update include files and add dir for example lib
- Move `struct vaccel_arg` definition
- Add Documentation

ananos commented on docs/vaccel_args.md:
````diff
@@ -0,0 +1,161 @@
+
+vaccel_add_deser_arg(read, &mystruct, 0, ser_mystruct);
+```
+Of cource, it must be emphasized that **it's up to the user to free the memory that was allocated by the serializer** 
````
choose consistent notation (serialize or serialise)

ananos commented on docs/vaccel_args.md:
````diff
@@ -0,0 +1,161 @@
+vaccel_expect_ser_arg(write, &out, sizeof(out));
+```
+#### Non-serialised
+Similarly, we call the corresponding function for non-serialised arguments. However we do not define any buffer, because the data there will be meaningless after the operation return, since it won't have been deserialised yet. Later (after calling the operation), we can ask for the output using their indeces, along with a deserializer function.
````
same as above (serialize or serialise?)

ananos commented on docs/vaccel_args.md:
````diff
@@ -0,0 +1,161 @@
+  ...
+}
+```
+**The memory allocated by the deserializer must be freed.**
````
same as above (serialize)

ananos commented on docs/vaccel_args.md:
````diff
@@ -0,0 +1,161 @@
+vaccel_write_ser_arg(write->list, 0, &response);
+...
+```
+#### Non-serialied
````
non-serialised (Typo)

ananos commented on docs/vaccel_args.md:
````diff
@@ -0,0 +1,161 @@
+vaccel_write_deser_arg(write->list, 1, &out, ser_mystruct);
+...
+```
+Since the serializer allocates memory, **don't forget to clean this memory later**.
````
same as above (notation)

ananos commented on examples/mytestlib.c:
```diff
@@ -22,23 +22,91 @@
 	printf("I got nr_in: %ld, nr_out: %ld\n", nr_in, nr_out);
-	printf("I got this input: %d\n", a);
-	sprintf(output[0].buf, "I got this input: %d\n", a);
-	output[0].len = strlen(output[0].buf);
```
since we have changed the way the library works (we return an integer, not a string) we should change the `exec.c`/`exec_with_res.c`/`exec_generic.c` example files that we use that, otherwise tests will fail.

ananos reviewed (changes requested):
LGTM, just make sure the library is consistent with the examples (all `exec` variants)

cmainas reviewed (changes requested):
Thank you a lot for this PR. We definitely needed to organize the handling of arguments. Overall, the PR looks good, but there are a few comments from my side. We could ignore some, but there are a few things we need to take care of.

We need to consider:
1. The memory allocation and indexing in arrays. We need to make sure that the memory allocation is successful and we do not corrupt any memory.
2. The documentation can get improved. We can use a high level to low level structure to explain how the arguments work and how the users can use them and what they need to implement. For instance, it was not clear how the (de)serialization should happen. If I understand correctly this depends on the application, but we should mention it.
3. We should add some information about argtypes.
4. Some files are missing license
5. (Optional) It would be nice, if we could use a specific code style. In general, in the past we used Linux kernel's coding style. There is also a perl script (checkpatch.pl) that can be used. FOr more info check here: https://www.kernel.org/doc/html/v4.10/process/coding-style.html

Non-critical (can get ignored):
1. The naming of some functions could be improved. For instance deser might refer to deserialization and not to unserialized args.
2. The vaccel_*_args functions could be changed from void to int, in order to allow error checking and handling for the caller.
3. The `vaccel_expect_*_args` could return an identifier that can be later used from `vaccel_extract_*_arg` to retrieve the respective argument. Please correct me, if I misunderstood the use of these functions.
4. We might want to add also a function that takes care of removing an argument from args_list, along with functions that add args.
5. We could use a different data structure for `vaccel_arg` in `vaccel_arg_list`. If I understand correctly we are currently using an array and we reallocate when adding a new arg. However, realloc is quite slow and calling it often might create some performance issues. Alternatives could be: a) a linked list, b) an array we defined size -- the size could be an argument in `vaccel_args_init`.
6. It would be nice if we can also add a few comments in the code here and there. People tend to not read the documentation, but they might read the comments.

ilagomatis commented on examples/exec_with_res.c:
```diff
@@ -135,9 +135,10 @@ int main(int argc, char *argv[])
 	input = 10; /* some random input value */
 	struct vaccel_arg read[1] = {
-		{.size = sizeof(input), .buf = &input}};
+		{.size = sizeof(input), .buf = &input, .argtype = 42}
```
Thank you for the comments. Regarding argtypes, I would like to ask if there's any policy that I should follow. I mean, the purpose for which the argtypes were introduced in vaccel_arg, isn't it to maintain a "list" of predefined serializers / deserializers? So that when a user adds a known arg-type, the helper functions may use the predefined serializers?

ananos replied:
these arbitrary values -- this field is actually a placeholder for a forthcoming PR that would allow two things: (a) generic ser/deser functions embedded in vaccelrt, (b) vAccel resources to be used as arguments (read/write) for vAccel plugins. For now we just add it and these values are just for debug (we expect to see them through various debug snippets throughout the code). just ignore for now.

ilagomatis commented on src/vaccel_args.c:
```diff
@@ -0,0 +1,133 @@
+        (struct vaccel_arg_list*)malloc(sizeof(struct vaccel_arg_list));
+
+    arg_list->size = 0;
+    arg_list->list = (struct vaccel_arg*)malloc(0);
```
We do that because we later call realloc() on arg_list->list. Realloc() requires that the memory area that is about to modify, has been previously allocated with malloc or equivalent

ilagomatis commented on src/vaccel_args.c:
```diff
@@ -0,0 +1,133 @@
+    void* (*serializer)(void*, uint32_t*))
+{
+    args->size++;
+    args->list = realloc(args->list, 
```
Generally, the helper functions don't return any value. Maybe I should modify them to return an integer? In order to return VACCEL_OK if everything is ok etc

cmainas replied:
Yeah, that sounds a good idea. However, we still need to check the memory allocation functions.

ananos commented on examples/exec_with_res.c:
```diff
@@ -135,9 +135,10 @@ int main(int argc, char *argv[])
+	};
 	struct vaccel_arg write[1] = {
-		{.size = sizeof(out_text), .buf = out_text},
+		{.size = sizeof(out_text), .buf = out_text, .argtype = 52},
```
same as above

ilagomatis commented on src/include/vaccel_args.h:
```diff
@@ -0,0 +1,71 @@
+    uint32_t size
+);
+
+void vaccel_expect_deser_arg(
```
I was thinking about it, too. I 'll change this. Something else I would like to ask if those names are considered too long. Maybe I can find some shorter ones

cmainas replied:
I think that the length of the name is ok and it might not be easy to shorten it. However, if you have any good idea, it would be welcome.

ilagomatis commented:
Hello, I added the following changes after your review:

1. vaccel_arg array is now allocated in the beginning providing a size in vaccel_args_init(), so that we can avoid using realloc()
2. I introduced a new mechanism of keeping track of arguments with allocated space so that the space can be freed in vaccel_delete_args(), and not let the user do it manually. It was checked with Valgrind and there are no leaks both in local and remote execution (using vaccelrt-agent, vsock-plugin etc)
3. Changed Naming
4. Checking results of malloc() (not in examples)
5. Add some comments
6. Helper functions now return a status (int)
7. Add Licence 

Also,
1. I tried to follow the same coding style as in vAccel. If you find any point that I didn't please correct me. 
2. Regarding idx and boundaries, I only check if idx >= 0 since we don't know the size of the arg-list at that point
3. Regarding vaccel_expect_*() returning an identifier, it would be a good idea. However, now they return a status. Do you think it would be better if we changed that?

cmainas reviewed (approved):
Thank you for updating the PR. It looks ok.

cmainas commented on examples/CMakeLists.txt:
Maybe, we could add `exec_helpers` and `exec_helpers_nonser` as binaries too.

ilagomatis commented on examples/mytestlib.c:
```diff
@@ -22,23 +22,107 @@
+}
+
+/* Deserializer function for `struct mydata` */
+void* deser(void* buf, uint32_t bytes){
```
`bytes` is unused in this case, since `deser` (deserializer of `struct mydata`) has access on the length of the array by reading the first 4 bytes of  the the serialized data ('buf' parameter). However, if the serialized data didn't contain the size, the deserializer wouldn't have access to it, so it could be provided as a parameter.

cmainas replied:
I see, that is fine, we can simply use a compiler attribute, in order to remove the compiler warning
```
--- a/examples/mytestlib.c
+++ b/examples/mytestlib.c
@@ -48,7 +48,7 @@ void* ser(void* buf, uint32_t* bytes){
 }
 
 /* Deserializer function for `struct mydata` */
-void* deser(void* buf, uint32_t bytes){
+void* deser(void* buf, uint32_t __attribute__((unused)) bytes){
        int *ser_buf = (int*)buf;
        int size = ser_buf[0];
```

ananos reviewed (approved):
Looks good! thanks @ilagomatis!
